src/session.rs

```diff
@@ -0,0 +1,159 @@
+/// NOTE: the passed in `key` and `val` are only availabe during the
+/// callback invocation.
+/// NOTE: callbacks used in several sessions via a common config
+/// must be implemented thread-safe.
+pub type rustls_session_store_put_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_session_store_userdata,
+        key: *const rustls_slice_bytes,
+        val: *const rustls_slice_bytes,
+    ) -> c_int,
+>;
+
+pub(crate) type SessionStorePutCallback = unsafe extern "C" fn(
+    userdata: rustls_session_store_userdata,
+    key: *const rustls_slice_bytes,
+    val: *const rustls_slice_bytes,
+) -> c_int;
+
+pub(crate) struct SessionStoreBroker {
+    pub userdata: rustls_session_store_userdata,
+    pub get_cb: SessionStoreGetCallback,
+    pub put_cb: SessionStorePutCallback,
+}
+
+impl SessionStoreBroker {
+    pub fn new(
+        userdata: rustls_session_store_userdata,
+        get_cb: SessionStoreGetCallback,
+        put_cb: SessionStorePutCallback,
+    ) -> Self {
+        SessionStoreBroker {
+            userdata,
+            get_cb,
+            put_cb,
+        }
+    }
+
+    fn retrieve(&self, key: &[u8], remove: bool) -> Option<Vec<u8>> {
+        let key: rustls_slice_bytes = key.into();
+        // TODO: we need a buffer where th client can store the retrieved
+        // session value. What size should it have? 10k seems excessive...
+        let mut data: Vec<u8> = vec![0 as u8; 10 * 1024];
+        let buffer = data.as_mut_slice();
+        let mut out_n: size_t = 0;
+        unsafe {
+            let cb = self.get_cb;
+            if cb(
+                self.userdata,
+                &key,
+                buffer.as_mut_ptr(),
+                buffer.len(),
+                remove as c_int,
+                &mut out_n,
+            ) != 0
+            {
+                data.set_len(out_n);
+                return Some(data);
+            }
+            None
+        }
+    }
+
+    fn store(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
+        let key: rustls_slice_bytes = key.as_slice().into();
+        let value: rustls_slice_bytes = value.as_slice().into();
+        unsafe {
+            let cb = self.put_cb;
+            cb(self.userdata, &key, &value) != 0
+        }
+    }
+}
+
+impl rustls::StoresServerSessions for SessionStoreBroker {
+    fn put(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
+        self.store(key, value)
+    }
+
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        return self.retrieve(key, false);
+    }
+
+    fn take(&self, key: &[u8]) -> Option<Vec<u8>> {
+        return self.retrieve(key, true);
+    }
+}
+
+impl rustls::StoresClientSessions for SessionStoreBroker {
+    fn put(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
+        self.store(key, value)
+    }
+
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        return self.retrieve(key, false);
+    }
+}
+
+unsafe impl Sync for SessionStoreBroker {}
+unsafe impl Send for SessionStoreBroker {}
```

Comment:
We should have a comment documenting why we consider this `Sync` and `Send` (even if the answer is just "we have documented that the C code needs to be threadsafe.")